src/journal/fsprg.c

```diff
@@ -96,119 +104,187 @@ _pure_ static uint64_t uint64_import(const void *buf, size_t buflen) {
-                gcry_md_putc(hd2, (ctr >> 24) & 0xff);
-                gcry_md_putc(hd2, (ctr >> 16) & 0xff);
-                gcry_md_putc(hd2, (ctr >>  8) & 0xff);
-                gcry_md_putc(hd2, (ctr >>  0) & 0xff);
-                gcry_md_final(hd2);
-                cpylen = (buflen < olen) ? buflen : olen;
-                memcpy(buf, gcry_md_read(hd2, RND_HASH), cpylen);
-                gcry_md_close(hd2);
+                if (EVP_MD_CTX_copy_ex(ctx2, ctx1) <=0)
+                        return -EIO;
+                if (EVP_DigestUpdate(ctx2, ((uint8_t*) &ctr + 3), sizeof(uint8_t)) <= 0)
+                        return -EIO;
+                if (EVP_DigestUpdate(ctx2, ((uint8_t*) &ctr + 2), sizeof(uint8_t)) <= 0)
+                        return -EIO;
+                if (EVP_DigestUpdate(ctx2, ((uint8_t*) &ctr + 1), sizeof(uint8_t)) <= 0)
+                        return -EIO;
+                if (EVP_DigestUpdate(ctx2, ((uint8_t*) &ctr), sizeof(uint8_t)) <= 0)
+                        return -EIO;
+                if (EVP_DigestFinal_ex(ctx2, md, NULL) <= 0)
+                        return -EIO;
+                cpylen = MIN(buflen, olen);
+                memcpy(buf, md, cpylen);
                 buf += cpylen;
                 buflen -= cpylen;
         }
-        gcry_md_close(hd);
+
+        return 0;
 }
 
 /* deterministically generate from seed/idx a prime of length `bits' that is 3 (mod 4) */
-static gcry_mpi_t genprime3mod4(int bits, const void *seed, size_t seedlen, uint32_t idx) {
+static BIGNUM* genprime3mod4(int bits, const void *seed, size_t seedlen, uint32_t idx) {
+        _cleanup_(BN_CTX_freep) BN_CTX *ctx = NULL;
         size_t buflen = bits / 8;
         uint8_t buf[buflen];
-        gcry_mpi_t p;
+        uint8_t buf2[buflen];
+        memset(buf2, 0, buflen);
+        BIGNUM *p;
 
         assert(bits % 8 == 0);
         assert(buflen > 0);
 
-        det_randomize(buf, buflen, seed, seedlen, idx);
+        ctx = BN_CTX_new();
+        if (!ctx)
+                /* Likely out of memory. */
+                return NULL;
+
+        if (det_randomize(buf, buflen, seed, seedlen, idx) < 0)
+                return NULL;
+
         buf[0] |= 0xc0; /* set upper two bits, so that n=pq has maximum size */
         buf[buflen - 1] |= 0x03; /* set lower two bits, to have result 3 (mod 4) */
 
         p = mpi_import(buf, buflen);
-        while (gcry_prime_check(p, 0))
-                gcry_mpi_add_ui(p, p, 4);
+        while (BN_is_prime_ex(p, BN_prime_checks, ctx, NULL) == 0)
+                assert(BN_add_word(p, 4) == 1);
 
         return p;
 }
 
 /* deterministically generate from seed/idx a quadratic residue (mod n) */
-static gcry_mpi_t gensquare(const gcry_mpi_t n, const void *seed, size_t seedlen, uint32_t idx, unsigned secpar) {
+static BIGNUM* gensquare(const BIGNUM *n, const void *seed, size_t seedlen, uint32_t idx, unsigned secpar) {
         size_t buflen = secpar / 8;
         uint8_t buf[buflen];
```

Comment:
@poettering should we remove these VLAs here and elsewhere in this file? not introduced in this PR, but might be worth fixing?

Reply:
I think this use of VLA is appropriate. The size is limited, and the VLA is more efficient than a malloc and nicer to use than explicit alloca. But if anything, this can be discussed separately.